routes/adjustmentMethods/EToAdjustmentMethod.ts

```diff
@@ -0,0 +1,224 @@
+import * as SunCalc from "suncalc";
+import * as moment from "moment";
+import { AdjustmentMethod, AdjustmentMethodResponse, AdjustmentOptions } from "./AdjustmentMethod";
+import { GeoCoordinates, WateringData, WeatherProviderId } from "../../types";
+import { WeatherProvider } from "../weatherProviders/WeatherProvider";
+
+
+/**
+ * Calculates how much watering should be scaled based on weather and adjustment options by comparing the recent
+ * potential ETo to the baseline potential ETo that the watering program was designed for.
+ */
+async function calculateEToWateringScale(
+	adjustmentOptions: EToScalingAdjustmentOptions,
+	wateringData: WateringData | undefined,
+	coordinates: GeoCoordinates,
+	weatherProvider: WeatherProvider
+): Promise< AdjustmentMethodResponse > {
+
+	// Temporarily disabled since OWM forecast data is checking if rain is forecasted for 3 hours in the future.
+	/*
+	if ( wateringData && wateringData.raining ) {
+		return {
+			scale: 0,
+			rawData: { raining: 1 }
+		}
+	}
+	 */
+
+	// This will throw an error message if ETo data cannot be retrieved.
+	const etoData: EToData = await weatherProvider.getEToData( coordinates );
+
+	let baseETo: number;
+	// Default elevation is based on data from https://www.pnas.org/content/95/24/14009.
+	let elevation = 600;
+
+	if ( adjustmentOptions && "baseETo" in adjustmentOptions ) {
+		baseETo = adjustmentOptions.baseETo
+	} else {
+		throw "A baseline potential ETo must be provided.";
+	}
+
+	if ( adjustmentOptions && "elevation" in adjustmentOptions ) {
+		elevation = adjustmentOptions.elevation;
+	}
+
+	const eto: number = calculateETo( etoData, elevation, coordinates );
+
+	const scale =  Math.floor( Math.min( Math.max( 0, ( eto - etoData.precip ) / baseETo * 100 ), 200 ) );
+	return {
+		scale: scale,
+		rawData: {
+			eto: Math.round( eto * 1000) / 1000,
+			radiation: Math.round( etoData.solarRadiation * 100) / 100
```

Comment:
Will we need to worry about how much these parameters will increase the size of the response?

Reply:
Yes, we do need to worry about buffer size. If you ever want to see some maximum efficiency code then take a look at Ray's Firmware. He eeks out every bit going. From memory Firmware uses either 128 or 256 bytes for the buffer depending on the hardware. That is why Zimmerman uses "h", "t", "p" as the pass parameters :) I would suggest the same here. By changing "rawData" and "raining" to "adj" and "r" then I think we stay aligned to Firmware and get the necessary bytes for the additional fields.

Reply:
I did some digging, and it looks like the smallest buffer used since version 2.1.0 of the firmware (the initial commit on GitHub) [was 850 bytes](https://github.com/OpenSprinkler/OpenSprinkler-Firmware/blob/f76984d7b05063aa9768df9192b45a1de4d45835/defines.h#L276), so we should plenty of room.

Do you by chance know if the buffer was even smaller in older firmware versions (and if those versions are even supported anymore)? If it has previously been <=256 bytes and HTTP headers are included in the buffer (not just the body), we already exceeded the buffer size and may need to remove or trim `rawData`

Reply:
Not just EtherBuffer for the entire packet. Check Firmware/weather.cpp for `tmp_buffer` that is either 128 or 256 bytes depending on hardware. Each parameter is copied into that field. So we need rawData to fit in that as well. Might not be a problem but better to know the limits.

Reply:
I hadn't noticed that limitation, thanks for the heads up!

Reply:
Pete is correct that currently each parameter is copied into a tmp_buffer which is 255 in size for OS 2.3, 3.0 and ospi. In fact, 255 is probably not a problem any ways, because all other parameters are of integer values so they are copied to tmp_buffer for conversion to integer. If a parameter is a string that does not need to be converted to integer, it does not need to be copied out to tmp_buffer, hence the 255 limit won't matter. The Ether buffer itself is at least 1400 bytes for OS 2.3, 3.0 and ospi. Any version prior to OS 2.3 will not receive new firmware updates.